joint_state_controller/src/joint_command_controller.cpp

```diff
@@ -0,0 +1,72 @@
+/**
+Software License Agreement (proprietary)
+
+\file      joint_command_controller.cpp
+\authors   Peiyi Chen <pchen@clearpath.ai>
+\copyright Copyright (c) 2022, Clearpath Robotics, Inc., All rights reserved.
+
+Redistribution and use in source and binary forms, with or without modification, is not permitted without the
+express permission of Clearpath Robotics.
+*/
+#include <controller_interface/internal/robothw_interfaces.h>
+#include "joint_state_controller/joint_command_controller.h"
+
+namespace joint_command_controller
+{
+
+bool JointCommandController::init(hardware_interface::RobotHW* hw, ros::NodeHandle& nh, ros::NodeHandle& controller_nh)
+{
+  controller_nh.param<double>("publish_rate", publish_rate_, publish_rate_);
+
+  cmd_pub_ = nh.advertise<sensor_msgs::JointState>("joint_commands", 10);
+
+  bool result = getJointHandles<hardware_interface::PositionJointInterface>(hw, pos_handles_) &&
+                getJointHandles<hardware_interface::VelocityJointInterface>(hw, vel_handles_) &&
+                getJointHandles<hardware_interface::EffortJointInterface>(hw, eff_handles_);
+
+  // The JointCommandInterface uses a HardwareResourceManager base class that auto claims resources when calling
+  // getHandle(). This causes resource conflicts, so it is important that the resources are unclaimed here.
+  controller_interface::internal::clearClaims<hardware_interface::PositionJointInterface,
+                                              hardware_interface::VelocityJointInterface,
+                                              hardware_interface::EffortJointInterface>(hw);
+
+  return result;
+}
+
+
+void JointCommandController::update(const ros::Time& time, const ros::Duration& period)
+{
+  auto populate_command = [&](const JointHandleMap& handles,
+                              std::vector<std::string>& joint_names,
+                              std::vector<double>& joint_cmds,
+                              std::vector<double>& unused1,
+                              std::vector<double>& unused2)
+  {
+    for (const auto [name, handle] : handles)
+    {
+      joint_names.push_back(name);
+      joint_cmds.push_back(handle.getCommand());
+      unused1.push_back(0.0);
+      unused2.push_back(0.0);
+    }
+  };
+
+  if (publish_rate_ <= 0.0 || (time - last_publish_time_).toSec() < (1.0 / publish_rate_))
```

Comment:
Shouldn't the `publish_rate_` enforce this already? why do we need this extra check? And `last_publish_time_` never seems to get written, making this check a no-op?

Reply:
Yea I messed up and forgot to update the `last_publish_time_`. 
We need the first check to ensure we don't divide by zero in the second half. If you mean why we need this at all, it's because the `update()` function is called at the rate of the controller_manager which is not necessarily the desired publish rate here (i.e. we may want to throttle it further).

Reply:
Yes, that was my question, throttling it down makes sense. Wondering if it would be better to throttle with integers and publish each update, every two, or every three etc... that avoids aliasing, but lets worry about that if it ever becomes a problem. :+1: